Added scaling capability to TilemapLayer

mickez opened the pull request:
See feature request  #1603 and #735

Usage is as following:
```
var layer = map.createLayer('layerName');
layer.setScale(4);
```

A common use case would be to make sure your pixel art game renders crisp at high resolution, without having to pre-process all files (which would cause an undesirable performance hit anyway).

JSFiddle example! http://jsfiddle.net/mickez/zp3f33hv/

it works by iterating through the underlying block data in the corresponding `map.layers` and set the `scale` property inherited from `Phaser.Image`. The rest of the code changes is to make sure the camera is moved accordingly and the correct size is rendered.

I'm unsure if all functions that would be affected by the maps size has been updated accordingly, but my tests have shown that rendering and collision works properly.

photonstorm commented:
Nice. I'll have a play with this later. I'd like to see a way to just over-ride layer.scale instead of adding another method to it (as I bet it would really mess things up if you set them both by mistake :))

mickez commented:
That would be ideal and more consistent!

It wouldn't really mess things up per se, it's just that if you set the scale without the added method the data used for collision doesn't get updated.

There is however one problem, if you create more than one TilemapLayer of a single actual layer, and scale one, but not the other, the collision of one of the TilemapLayer's will be messed up! The only real solution to this would be to make a copy of the data instead of using the current approach using a reference.

I did play around without adding a new method. The data used for collision (in arcade physics at least) is the left/top/right/bottom-properties in the Tile-class. If every instance of Tile has a reference to the TilemapLayer it belongs to it would be trivial to just do the calculation inside the get function. I opted against it in this PR as the Tile data is populated with the creation of the TileMap and not with the creation of the TilemapLayer and I didn't really want to re-structure too much of the code :).

mickez commented:
Another approach to solve both problems described above would be to calculate the scaling of the TilemapLayer inside the arcade physics collision checks itself. This would remove the need to edit any data (that is to just simply use the scale property) and in turn still have the ability to add two TilemapLayers which uses the same data

photonstorm commented:
For some reason your jsfiddle doesn't work in WebGL mode, but maybe that's just the build you ran off? I get nothing rendered, just a grey screen.

mickez commented:
Must be the build, if not, surely the character and text should have rendered! I'll test webgl later today with my commits rolled back
________________________________
Från: Richard Davey<mailto:notifications@github.com>
Skickat: ‎2015-‎02-‎08 22:28
Till: photonstorm/phaser<mailto:phaser@noreply.github.com>
Kopia: Michael<mailto:rehn.michael@outlook.com>
Ämne: Re: [phaser] Added scaling capability to TilemapLayer (#1605)

For some reason your jsfiddle doesn't work in WebGL mode, but maybe that's just the build you ran off? I get nothing rendered, just a grey screen.

---
Reply to this email directly or view it on GitHub:
https://github.com/photonstorm/phaser/pull/1605#issuecomment-73432956

mickez commented:
It was just cross-domain complaints! I changed the images to base64 encoded images so I could have them inline. The jsfiddle is now updated and works with WebGL: http://jsfiddle.net/mickez/zp3f33hv/1/

photonstorm commented:
Perfect, I didn't have time to investigate fully but thanks for confirming. Ok let's get this merged and see what happens!
